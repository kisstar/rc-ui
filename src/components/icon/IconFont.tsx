// Iconfont
//
// Manually forked from ant-design-icons/packages/icons-react/
// How it works：https://www.iconfont.cn/help/detail?spm=a313x.7781069.1998910419.d0091c141&helptype=code
//
// ant-design-icons is licensed MIT. https://github.com/ant-design/ant-design-icons

import React, { useContext } from 'react';
import classNames from 'classnames';
import { ConfigContext } from '../config-provider';

export interface CustomIconOptions {
  scriptUrl?: string | string[];
}

export interface IconFontProps<T = string> {
  type: T;
  className?: string;
  style?: React.CSSProperties;
  spin?: boolean;
  rotate?: number;
  onClick?: React.MouseEventHandler<SVGSVGElement>;
  prefixCls?: string;
}

const customCache = new Set<string>();

function isValidCustomScriptUrl(scriptUrl: string): boolean {
  return Boolean(typeof scriptUrl === 'string' && scriptUrl.length && !customCache.has(scriptUrl));
}

function createScriptUrlElements(scriptUrls: string[], index: number = 0): void {
  const currentScriptUrl = scriptUrls[index];
  if (isValidCustomScriptUrl(currentScriptUrl)) {
    const script = document.createElement('script');
    script.setAttribute('src', currentScriptUrl);
    script.setAttribute('data-namespace', currentScriptUrl);
    if (scriptUrls.length > index + 1) {
      script.onload = () => {
        createScriptUrlElements(scriptUrls, index + 1);
      };
      script.onerror = () => {
        createScriptUrlElements(scriptUrls, index + 1);
      };
    }
    customCache.add(currentScriptUrl);
    document.body.appendChild(script);
  }
}

export default function create<T>(options: CustomIconOptions = {}): React.FC<IconFontProps<T>> {
  const { scriptUrl = '' } = options;

  /**
   * DOM API required, make sure in browser environment.
   * The Custom Icon will create a <script/>
   * that loads SVG symbols and insert the SVG Element into the document body.
   */
  if (
    scriptUrl &&
    typeof document !== 'undefined' &&
    typeof window !== 'undefined' &&
    typeof document.createElement === 'function'
  ) {
    if (Array.isArray(scriptUrl)) {
      // 因为 IconFont 资源会把 svg 插入 before，所以前加载相同 type 会覆盖后加载
      // 倒叙插入，为了数组覆盖顺序
      createScriptUrlElements(scriptUrl.reverse());
    } else {
      createScriptUrlElements([scriptUrl]);
    }
  }

  const Iconfont: React.FC<IconFontProps<T>> = props => {
    const { type, style = {}, className, prefixCls: customizePrefixCls, spin, rotate } = props;
    const { getPrefixCls } = useContext(ConfigContext);
    const prefixCls = getPrefixCls('icon', customizePrefixCls);
    const classes = classNames(prefixCls, className, {
      [`${prefixCls}-spin`]: !!spin,
    });
    const svgStyle = rotate
      ? {
          msTransform: `rotate(${rotate}deg)`,
          transform: `rotate(${rotate}deg)`,
        }
      : {};
    const retStyle = Object.assign(svgStyle, style);

    return (
      <svg className={classes} style={retStyle} aria-hidden="true">
        <use xlinkHref={`#icon-${type}`} />
      </svg>
    );
  };

  Iconfont.displayName = 'Iconfont';

  return Iconfont;
}
