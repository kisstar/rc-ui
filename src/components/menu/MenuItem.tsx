import React, { useCallback, useContext } from 'react';
import classNames from 'classnames';
import MenuContext from './MenuContext';

export interface MenuItemProps {
  /** 唯一标志 */
  key: string;
  /** 是否禁用 */
  disabled?: boolean;
  style?: React.CSSProperties;
  className?: string;
  /** 菜单图标 */
  icon?: React.ReactNode;
  rootPrefixCls?: string;
  eventKey?: string;
}

export const MenuItem: React.FC<MenuItemProps> = props => {
  const { rootPrefixCls, className, style, disabled, icon, children, eventKey } = props;
  const { onSelect, selectedKey } = useContext(MenuContext);
  const classes = classNames(`${rootPrefixCls}-item`, className, {
    [`${rootPrefixCls}-item-disabled`]: disabled,
    [`${rootPrefixCls}-item-selected`]: eventKey === selectedKey,
  });
  const handleClick = useCallback(
    (e: React.MouseEvent<HTMLLIElement>) => {
      const info = {
        key: eventKey as string,
        domEvent: e,
      };
      if (!disabled) {
        onSelect!(info);
      }
    },
    [eventKey, disabled, onSelect],
  );
  const onKeyDown = useCallback(
    (e: React.KeyboardEvent<HTMLElement>): boolean | undefined => {
      const { keyCode } = e;
      if (keyCode === 13 && !disabled) {
        onSelect!({ key: eventKey as string, domEvent: e });
        return true;
      }
      return undefined;
    },
    [eventKey, disabled, onSelect],
  );

  return (
    <li
      role="menuitem"
      className={classes}
      style={style}
      onClick={handleClick}
      onKeyDown={onKeyDown}
    >
      {icon}
      {children}
    </li>
  );
};

MenuItem.displayName = 'MenuItem';

export default MenuItem;
