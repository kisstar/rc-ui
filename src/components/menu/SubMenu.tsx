import React, { useState, useContext } from 'react';
import classNames from 'classnames';
import MenuContext from './MenuContext';
import { MenuItemProps } from './MenuItem';

export interface SubMenuProps {
  /** 子菜单项值 */
  title: string;
  style?: React.CSSProperties;
  className?: string;
  /** 是否禁用 */
  disabled?: boolean;
  /** 菜单图标 */
  icon?: React.ReactNode;
  rootPrefixCls?: string;
  eventKey?: string;
}

interface SubMenuType extends React.FC<SubMenuProps> {
  mouseTimeout: any;
}

export const SubMenu: SubMenuType = props => {
  const { rootPrefixCls, style, className, title, disabled, icon, children } = props;
  const [isOpened, setIsOpened] = useState(false);
  const [isSelected, setIsSelected] = useState(false);
  const { mode, selectedKey } = useContext(MenuContext);
  const classes = classNames(`${rootPrefixCls}-submenu`, className, {
    [`${rootPrefixCls}-submenu-disabled`]: disabled,
    [`${rootPrefixCls}-submenu-selected`]: isSelected,
    [`${rootPrefixCls}-submenu-open`]: isOpened,
  });

  const handleClick = () => {
    if (disabled || mode === 'horizontal' || mode === 'responsive') return;
    setIsOpened(!isOpened);
  };

  const onKeyDown = () => {
    if (disabled || mode === 'horizontal' || mode === 'responsive') return;
    setIsOpened(!isOpened);
  };

  const handleMouseEnter = () => {
    if (disabled || mode === 'vertical') return;
    clearTimeout(SubMenu.mouseTimeout);
    SubMenu.mouseTimeout = setTimeout(() => {
      setIsOpened(true);
    }, 100);
  };

  const handleMouseLeave = () => {
    if (disabled || mode === 'vertical') return;
    clearTimeout(SubMenu.mouseTimeout);
    SubMenu.mouseTimeout = setTimeout(() => {
      setIsOpened(false);
    }, 100);
  };

  const renderChild = () => {
    const keys: string[] = [];

    const retChild = (
      <ul
        role="menu"
        onMouseEnter={handleMouseEnter}
        onMouseLeave={handleMouseLeave}
        className={`${rootPrefixCls} ${rootPrefixCls}-vertical`}
      >
        {React.Children.map(children, child => {
          const childElement = child as React.FunctionComponentElement<MenuItemProps>;
          const { type, key } = childElement;
          const { displayName } = type;
          const strKey = key as string;

          keys.push(strKey);

          if (displayName === 'MenuItem') {
            return React.cloneElement(childElement, {
              eventKey: strKey,
              rootPrefixCls,
            });
          }

          throw Error("SubMenu's child can only be MenuItem");
        })}
      </ul>
    );

    const newSelect = keys.includes(selectedKey);

    if (newSelect !== isSelected) {
      setIsSelected(newSelect);
    }

    if (!isOpened) {
      return null;
    }

    return retChild;
  };

  return (
    <li style={style} className={classes}>
      <div
        role="button"
        tabIndex={-1}
        className={`${rootPrefixCls}-submenu-title`}
        onClick={handleClick}
        onKeyDown={onKeyDown}
        onMouseEnter={handleMouseEnter}
        onMouseLeave={handleMouseLeave}
      >
        <span>
          {icon}
          {title}
        </span>
        <i className={`${rootPrefixCls}-submenu-arrow`} />
      </div>
      {renderChild()}
    </li>
  );
};

SubMenu.displayName = 'SubMenu';
SubMenu.mouseTimeout = 0;

export default SubMenu;
