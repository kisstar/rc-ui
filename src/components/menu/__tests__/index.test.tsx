import React from 'react';
import { render, fireEvent, wait } from '@testing-library/react';
import { getStyleStr, mountTest, expectToThrow } from '../../../tests';
import { Menu, MenuProps } from '..';
import { SubMenu } from '../SubMenu';
import { MenuItem } from '../MenuItem';
import { Icon } from '../../icon';

const genMenu = (props?: MenuProps) => {
  return (
    <Menu data-testid="menu" defaultSelectedKey="1" {...props}>
      <MenuItem key="1">Active</MenuItem>
      <MenuItem key="2">Normal</MenuItem>
      <MenuItem key="3" disabled>
        Disabled
      </MenuItem>
      <SubMenu title="SubItems">
        <MenuItem key="4">First SubItem</MenuItem>
        <MenuItem key="5">Second SubItem</MenuItem>
      </SubMenu>
    </Menu>
  );
};
const prefixCls = 'ks-menu';
const activeClass = `${prefixCls}-item-selected`;

describe('Menu', () => {
  it('renders correctly', () => {
    expect(genMenu()).toMatchSnapshot();
  });

  mountTest(() => genMenu());

  it('should render correct Menu and MenuItem based on the specified props', () => {
    const wrapper = render(genMenu());
    const menuElement = wrapper.getByRole('menu');
    const activeElement = wrapper.getByText('Active');
    const disabledElement = wrapper.getByText('Disabled');

    expect(menuElement).toBeInTheDocument();
    expect(menuElement).toHaveClass(`${prefixCls} ${prefixCls}-horizontal`);
    expect(activeElement).toHaveClass(activeClass);
    expect(disabledElement).toHaveClass(`${prefixCls}-item-disabled`);
    expect(menuElement.getElementsByTagName('li').length).toEqual(4);
    expect(menuElement.getElementsByTagName('li')[0]).toHaveClass(`${prefixCls}-item`);
  });

  it("Menu's child should only be SubMenu or MenuItem", () => {
    expectToThrow(() => {
      render(
        <Menu data-testid="menu">
          <li>Menu Item</li>
        </Menu>,
      );
    });
  });

  it("SubMenu's child should only be MenuItem", () => {
    expectToThrow(() => {
      render(
        <Menu data-testid="menu">
          <SubMenu title="SubItems">
            <li>Menu Item</li>
          </SubMenu>
        </Menu>,
      );
    });
  });

  it('should render vertical mode when mode is set to vertical', () => {
    const wrapper = render(genMenu({ mode: 'vertical' }));
    const menuElement = wrapper.getByRole('menu');

    expect(menuElement).toBeInTheDocument();
    expect(menuElement).toHaveClass(`${prefixCls} ${prefixCls}-vertical`);
  });

  it('click items should change active and call the right callback', () => {
    const testProps = { onSelect: jest.fn() };
    const wrapper = render(genMenu(testProps));
    const activeElement = wrapper.getByText('Active');
    const disabledElement = wrapper.getByText('Disabled');
    const normalItem = wrapper.getByText('Normal');
    const args = expect.objectContaining({
      key: '2',
      domEvent: expect.any(Object),
    });

    fireEvent.click(normalItem);
    expect(normalItem).toHaveClass(activeClass);
    expect(activeElement).not.toHaveClass(activeClass);
    expect(testProps.onSelect).toHaveBeenCalledWith(args);
    testProps.onSelect.mockRestore();
    fireEvent.click(disabledElement);
    expect(normalItem).toHaveClass(activeClass);
    expect(disabledElement).not.toHaveClass(activeClass);
    expect(testProps.onSelect).not.toHaveBeenCalledWith(expect.any(Object));
  });

  it('should trigger mouse click when enter key down', () => {
    const testProps = { onSelect: jest.fn() };
    const wrapper = render(genMenu(testProps));
    const normalItem = wrapper.getByText('Normal');
    const disabledElement = wrapper.getByText('Disabled');
    const args = expect.objectContaining({
      key: '2',
      domEvent: expect.any(Object),
    });

    fireEvent.keyDown(normalItem, { key: 'Enter', keyCode: 13 });
    expect(testProps.onSelect).toHaveBeenCalledWith(args);
    testProps.onSelect.mockRestore();
    fireEvent.keyDown(disabledElement, { key: 'Enter', keyCode: 13 });
    expect(testProps.onSelect).not.toHaveBeenCalledWith(args);
  });

  it('should show dropdown items when hover on subMenu', async () => {
    const testProps = { onSelect: jest.fn() };
    const wrapper = render(genMenu(testProps));
    const subElement = wrapper.getByText('SubItems');

    expect(wrapper.queryByText('First SubItem')).not.toBeInTheDocument();
    fireEvent.mouseEnter(subElement);
    await wait(() => {
      // notice that setTimeout was used in the component before
      expect(wrapper.queryByText('First SubItem')).toBeInTheDocument();
    });
    fireEvent.click(wrapper.getByText('First SubItem'));
    expect(testProps.onSelect).toHaveBeenCalledWith({
      key: '4',
      domEvent: expect.any(Object),
    });
    fireEvent.mouseLeave(subElement);
    await wait(() => {
      expect(wrapper.queryByText('First SubItem')).not.toBeInTheDocument();
    });
  });

  it('should show dropdown items when click on subMenu with mode vertical', async () => {
    const testProps: MenuProps = {
      mode: 'vertical',
      onSelect: jest.fn(),
    };
    const wrapper = render(genMenu(testProps));
    const subElement = wrapper.getByText('SubItems');

    expect(wrapper.queryByText('Second SubItem')).not.toBeInTheDocument();
    fireEvent.mouseEnter(subElement);
    await wait(() => {
      expect(wrapper.queryByText('Second SubItem')).not.toBeInTheDocument();
    });

    fireEvent.click(subElement);
    await wait(() => {
      expect(wrapper.queryByText('Second SubItem')).toBeInTheDocument();
    });

    fireEvent.click(wrapper.getByText('Second SubItem'));
    expect(testProps.onSelect).toHaveBeenCalledWith({
      key: '5',
      domEvent: expect.any(Object),
    });

    fireEvent.click(subElement);
    await wait(() => {
      expect(wrapper.queryByText('Second SubItem')).not.toBeInTheDocument();
    });
  });

  it('should show dropdown items when enter key down with mode vertical', async () => {
    const testProps: MenuProps = {
      mode: 'vertical',
      onSelect: jest.fn(),
    };
    const wrapper = render(genMenu(testProps));
    const subElement = wrapper.getByText('SubItems');

    expect(wrapper.queryByText('Second SubItem')).not.toBeInTheDocument();
    fireEvent.keyDown(subElement, { key: 'Enter', keyCode: 13 });
    await wait(() => {
      expect(wrapper.queryByText('Second SubItem')).toBeInTheDocument();
    });

    fireEvent.keyDown(subElement, { key: 'Enter', keyCode: 13 });
    await wait(() => {
      expect(wrapper.queryByText('Second SubItem')).not.toBeInTheDocument();
    });
  });

  it('MenuItem should render icon and icon should be the first child when icon exists', () => {
    const wrapper = render(
      <Menu>
        <Menu.Item key="link" icon={<Icon data-testid="link" icon="link" />}>
          Navigation One
        </Menu.Item>
      </Menu>,
    );
    expect(expect(wrapper.getByTestId('link')).toBeInTheDocument());
  });

  it('should add cross styles to the correct dom', () => {
    const testClassName = 'test';
    const testStyle = { backgroundColor: '#f8f9fa' };
    const wrapper = render(
      genMenu({
        mode: 'responsive',
        className: testClassName,
        style: testStyle,
      }),
    );
    const navEle = wrapper.getByTestId('menu');
    const ele = wrapper.getByRole('menu');

    expect(navEle).toBeInTheDocument();
    expect(navEle).toHaveClass(testClassName);
    expect(navEle).toHaveStyle(getStyleStr(testStyle));
    expect(ele).toBeInTheDocument();
    expect(ele).not.toHaveClass(testClassName);
    expect(ele).not.toHaveStyle(getStyleStr(testStyle));
  });

  it('should display the DOM correctly in responsive mode', () => {
    const wrapper = render(genMenu({ mode: 'responsive' }));
    const navEle = wrapper.getByTestId('menu');
    const ele = wrapper.getByRole('menu');
    const buttonEle = navEle.firstElementChild as Element;

    expect(buttonEle).toBeInTheDocument();
    expect(buttonEle.tagName).toEqual('BUTTON');
    // expect(buttonEle).not.toBeVisible();
    // expect(getComputedStyle(buttonEle).display).toEqual('none');

    fireEvent.click(buttonEle);

    expect(ele).toHaveClass(`${prefixCls}-show`);
  });
});
