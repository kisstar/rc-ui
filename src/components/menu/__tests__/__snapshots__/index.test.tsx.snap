// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly 1`] = `
<Menu
  defaultSelectedKey="1"
  mode="horizontal"
  onSelect={[Function]}
>
  <MenuItem>
    Active
  </MenuItem>
  <MenuItem>
    Normal
  </MenuItem>
  <MenuItem
    disabled={true}
  >
    Disabled
  </MenuItem>
  <SubMenu
    title="SubItems"
  >
    <MenuItem>
      First SubItem
    </MenuItem>
    <MenuItem>
      Second SubItem
    </MenuItem>
  </SubMenu>
</Menu>
`;
